doc/coq.rst

```diff
@@ -0,0 +1,494 @@
+.. _coq:
+
+***
+Coq
+***
+
+
+Dune is able to build Coq projects and plugins with support for source file
+preprocessing, extraction and documenation.
+
+
+.. contents:: Table of Contents
+    :depth: 3
+
+.. _coq-theory:
+
+coq.theory
+----------
+
+A Coq project consists of Coq ``.v`` files and optional OCaml libraries linking
+to the Coq API (in which case we say the project is a *Coq plugin*). To enable
+Coq support in a Dune project, the language version should be selected in the
+``dune-project`` file. For example:
+
+.. code:: scheme
+
+    (using coq 0.3)
+
+This will enable support for the ``coq.theory`` stanza in the current project.
+If the language version is absent, Dune will automatically add this line with
+the latest Coq version to the project file once a ``(coq.theory ...)`` stanza is
+used anywhere.
+
+The basic form for defining Coq libraries is very similar to the OCaml form:
+
+.. code:: scheme
+
+    (coq.theory
+     (name <module_prefix>)
+     (package <package>)
+     (synopsis <text>)
+     (modules <ordered_set_lang>)
+     (libraries <ocaml_libraries>)
+     (flags <coq_flags>)
+     (mode <coq_native_mode>)
+     (theories <coq_theories>))
+
+The stanza will build all ``.v`` files on the given directory. The semantics of
+fields are:
+
+- ``<module_prefix>`` is a dot-separated list of valid Coq module names and
+  determines the module scope under which the theory is compiled [``-R``
+  option]. For example, if ``<module_prefix>`` is ``foo.Bar``, the theory
+  modules will be named as ``foo.Bar.module1``, ``foo.Bar.module2``, etc. Note
+  that modules in the same theory don't see the ``foo.Bar`` prefix in the same
+  way that OCaml ``wrapped`` libraries do. For compatibility reasons, the Coq
+  v1.0 installs a theory named ``foo.Bar`` under ``foo/Bar``. Also note that Coq
+  supports composing a module path from different theories, thus you can name a
+  theory ``foo.Bar`` and a second one ``foo.Baz``, and things will work
+  properly.
```

Comment:
"Coq v1.0" will be confusing for Coq users.  They expect something like "Coq v8.15".

I think it's a mistake to try to explain Coq concepts here--that should be in the Coq documentation only.

Reply:
It shouldn't be Coq v1.0 but rather Coq lang v1.0. We use versioning for the coq parts of dune so that files can cleanly reject unsupported features on older versions of dune.

Reply:
I've updated this.